### DISCUSSION TITLE
How do you generate tints and shades?

### DISCUSSION THREAD

**Discussion by binajmen** [tailwindlabs/tailwindcss#9044] [Help] [answered]
Hi 👋 

In my current project, I let a user define primary and secondary colors. It was easy to extend tailwind with these colors:
```js
module.exports = {
    extend: {
      colors: {
        // https://tailwindcss.com/docs/customizing-colors#using-css-variables
        primary: "rgb(var(--my-primary-color) / <alpha-value>)",
        secondary: "rgb(var(--my-secondary-color) / <alpha-value>)",
      },
    },
  },
};
```

I'd like to use these colors in different components. Let's take the example of a button. In my case, a button has three state: normal, hover, and focus. Therefore, similar to what I would do with a tailwind provided color, I would like to implement a button this way:

```js
<button className="bg-primary-500 hover:bg-primary-400 focus:bg-primary-700">...</button>
```

Where `bg-primary-500` would be the base color (for the sake of the example).

What is the correct way to generate these tints/shades starting from a dynamic color?

Any feedback would be greatly appreciated :)

**Comment by wongjn** [comment, marked as answer]
You could split the color CSS variables into components, say:

```css
:root {
  --primary-red: 255;
  --primary-green: 100;
  --primary-blue: 100;
  --secondary-red: 100;
  --secondary-green: 100;
  --secondary-blue: 255;
}
```

Then, in your Tailwind config, you could use `calc()` in the shades:

```js
module.exports = {
    extend: {
      colors: {
        primary: {
          // …
          400: 'rgb(calc(var(--primary-red) * <something>) calc(var(--primary-green) * <something>)
 calc(var(--primary-blue) * ) / <alpha-value>)',
          500: 'rgb(var(--primary-red) var(--primary-green) var(--primary-blue) / <alpha-value>)',
          600: 'rgb(calc(var(--primary-red) * <something>) calc(var(--primary-green) * <something>)
 calc(var(--primary-blue) * <something>) / <alpha-value>)',
          // …
        },
      },
    },
  },
};
```

Not sure what the best thing would be to put in `calc()` though. Perhaps using HSL color format would be easier, since you probably would only need to tweak the saturation and lightness components, and (for me anyway), the adjustments would seem more intuitive to think about.

**Reply by binajmen** [reply]
Oh, I just read about HSL thanks to your suggestion. Clearly, the third parameter – _lightness_ – does make sense in this specific use case! I'm wondering if I can simply do:
```
primary: {
  lighter: 'hsl(var(--primary-hue), var(--primary-saturation), calc(var(--primary-lightness) * 0.75))',
  base: 'hsl(var(--primary-hue), var(--primary-saturation), var(--primary-lightness))',
  ...
```

**Reply by wongjn** [reply]
Yup, that would work, don't forget the alpha component if you need it (using modern HSL syntax):

```js
lighter: 'hsl(var(--primary-hue) var(--primary-saturation) calc(var(--primary-lightness) * 0.75) / <alpha-value>)',
```

**Reply by binajmen** [reply]
Was testing it with success, thanks for the suggestion, clearly solve my problem :)
